src/target/breakpoints.c

```diff
@@ -287,60 +295,89 @@ static void breakpoint_free(struct target *target, struct breakpoint *breakpoint
 	if (!breakpoint)
 		return;
 
-	retval = target_remove_breakpoint(target, breakpoint);
+	retval = target_remove_breakpoint(breakpoint_target, breakpoint);
 
 	LOG_DEBUG("free BPID: %" PRIu32 " --> %d", breakpoint->unique_id, retval);
 	(*breakpoint_p) = breakpoint->next;
 	free(breakpoint->orig_instr);
 	free(breakpoint);
 }
 
-static int breakpoint_remove_internal(struct target *target, target_addr_t address)
-{
-	struct breakpoint *breakpoint = target->breakpoints;
-
-	while (breakpoint) {
-		if ((breakpoint->address == address) ||
-		    (breakpoint->address == 0 && breakpoint->asid == address))
-			break;
-		breakpoint = breakpoint->next;
-	}
-
-	if (breakpoint) {
-		breakpoint_free(target, breakpoint);
-		return 1;
-	} else {
-		if (!target->smp)
-			LOG_ERROR("no breakpoint at address " TARGET_ADDR_FMT " found", address);
-		return 0;
-	}
-}
-
 static void breakpoint_remove_all_internal(struct target *target)
 {
 	struct breakpoint *breakpoint = target->breakpoints;
 
 	while (breakpoint) {
 		struct breakpoint *tmp = breakpoint;
 		breakpoint = breakpoint->next;
-		breakpoint_free(target, tmp);
+		breakpoint_free(target, target, tmp);
 	}
 }
 
 void breakpoint_remove(struct target *target, target_addr_t address)
 {
 	if (target->smp) {
-		unsigned int num_breakpoints = 0;
+		unsigned int found = 0;
 		struct target_list *head;
+		/* Target where we found a software breakpoint. */
+		struct target *software_breakpoint_target = NULL;
+		struct breakpoint *software_breakpoint = NULL;
```

Comment:
If you like, these identifiers can be shortened without making them unclear:

```suggestion
		struct target *sw_breakpoint_target = NULL;
		struct breakpoint *sw_breakpoint = NULL;
```

Reply:
I don't think shorter is better in this case.

Reply:
Ok, let's leave it as is.